mozilla_version/balrog.py

```diff
@@ -0,0 +1,153 @@
+"""Defines characteristics of a Balrog release name.
+
+Balrog is the server that delivers Firefox and Thunderbird updates. Release names follow
+the pattern "{product}-{version}-build{build_number}"
+
+Examples:
+    .. code-block:: python
+
+        from mozilla_version.balrog import BalrogReleaseName
+
+        balrog_release = BalrogReleaseName.parse('firefox-60.0.1-build1')
+
+        balrog_release.product                 # firefox
+        balrog_release.version.major_number    # 60
+        str(balrog_release)                    # 'firefox-60.0.1-build1'
+
+        previous_release = BalrogReleaseName.parse('firefox-60.0-build2')
+        previous_release < balrog_release      # True
+
+        invalid = BalrogReleaseName.parse('60.0.1')           # raises PatternNotMatchedError
+        invalid = BalrogReleaseName.parse('firefox-60.0.1')   # raises PatternNotMatchedError
+
+        # Releases can be built thanks to version classes like FirefoxVersion
+        BalrogReleaseName('firefox', FirefoxVersion(60, 0, 1, 1))  # 'firefox-60.0.1-build1'
+
+"""
+
+import attr
+import re
+
+from mozilla_version.errors import PatternNotMatchedError
+from mozilla_version.parser import parse_and_construct_object, get_value_matched_by_regex
+from mozilla_version.gecko import (
+    GeckoVersion, FirefoxVersion, DeveditionVersion, FennecVersion, ThunderbirdVersion
+)
+
+
+_VALID_ENOUGH_BALROG_RELEASE_PATTERN = re.compile(r"""
+^(?P<product>[a-z]+)
+-(?P<major_number>\d+)
+\.(?P<minor_number>\d+)
+(\.(?P<patch_number>\d+))?
+(
+    (?P<is_nightly>a1)
+    |(?P<is_aurora_or_devedition>a2)
+    |b(?P<beta_number>\d+)
+    |(?P<is_esr>esr)
+)?
+-build(?P<build_number>\d+)
+$""", re.VERBOSE | re.IGNORECASE)
```

Comment:
I don't think you should try to parse the version here. I think you should just capture the entire version string (maybe even with `.*`) and then pass it to the parser for the version. And, if you do that, I don't think you need to make the version parsing logic public.

Reply:
Good thinking. Fixed!